src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+        self.rng: np.random.RandomState = np.random.RandomState(seed=rng_seed)
+
+        # generate gene names
+        self.all_gene_names = self._load_gene_names(annotation_file)
+        self.max_genes = max_genes
+        self.num_genes = len(self.all_gene_names)
+        self.to_be_used_gene_indices: List[int] = self.rng.choice(
+            np.arange(0, self.num_genes, dtype=np.int), size=self.max_genes, replace=False).tolist()
+        self.to_be_used_gene_names = [self.all_gene_names[j] for j in self.to_be_used_gene_indices]
+
+    def generate_synthetic_counts_and_alignment_tags(self,
+                                                     num_duplicates: int,
+                                                     num_missing_some_tags: int,
+                                                     num_multiple_gene_alignments: int,
+                                                     max_gene_hits_per_multiple_gene_alignments: int)\
+            -> SyntheticDataBundle:
+
+        # generate count matrix
+        count_matrix: np.ndarray = self._generate_random_count_matrix()
+
+        # generate primary alignment tags that produce count_matrix
+        primary_alignment_record_tags_set, row_index, col_index = \
+            self._generate_primary_alignment_record_bundle(count_matrix)
+        primary_alignment_record_tags_list = list(primary_alignment_record_tags_set)
+
+        # sanity check -- we require as many primary alignment records as the total counts
+        assert len(primary_alignment_record_tags_set) == np.sum(count_matrix)
+
+        # add duplicate records
+        duplicate_alignment_tags_list = self._generate_duplicate_alignment_tags(
+            num_duplicates, primary_alignment_record_tags_list)
+
+        # add records with missing tags
+        incomplete_alignment_tags_list: List[AlignmentRecordTags] = self._generate_incomplete_alignment_tags(
+            num_missing_some_tags)
+
+        # add records with multiple gene alignments
+        multiple_alignment_tags_list: List[List[AlignmentRecordTags]] = self._generate_multiple_gene_alignment_tags(
+            num_multiple_gene_alignments,
+            max_gene_hits_per_multiple_gene_alignments,
+            primary_alignment_record_tags_set)
+
+        return SyntheticDataBundle(count_matrix,
+                                   row_index,
+                                   col_index,
+                                   primary_alignment_record_tags_list,
+                                   duplicate_alignment_tags_list,
+                                   incomplete_alignment_tags_list,
+                                   multiple_alignment_tags_list)
+
+    def generate_synthetic_bam_and_counts_matrix(self,
+                                                 output_path: str,
+                                                 num_duplicates: int,
+                                                 num_missing_some_tags: int,
```

Comment:
Can we add a guarantee that each tag is missed at least once for testing purposes, or split this out into three parameters, one per missing tag?

Reply:
Done.